[SLM] UX improvement: HF download & presets

davidpissarra opened the pull request:
This PR proposes additions to the new compilation CLI to complete A3 and A4 from #1364:

- [A3] Download `mlc-chat-config.json` from HF to compile the model;
- [A4] Model compilation through presets.

cc @junrushao @tqchen

junrushao commented:
Hey @davidpissarra thanks for this nice PR!

I'd love to briefly bring up the change I introduced yesterday and discuss with you on potentially minimizing the metadata needed in compilation: https://github.com/mlc-ai/mlc-llm/pull/1375. To recap, the changes relevant to this PR is:

> Feeds only `model_config` field in `mlc-chat-config.json` to the
> compile interface to make sure the compilation is always reproducible
> depending on this single field plus "quantization";

It means that, in principal, `MODEL_PRESETS` should suffice to compile a model, meaning `HF_PRESET` may not be required.

To further discuss why we need `MODEL_PRESETS`, it is a set of weight-agnostic model configurations that could help us focus on the model per se, for example, say two finetunes of Llama2-7B exist but their architecture is exactly the same. In those cases, they could share the generated model lib without re-compilation, by looking up the names in `MODEL_PRESETS`

davidpissarra commented:
Hi @junrushao, thanks for the great explanation! To test my understanding about this item, let me ask you a quick question. Since we intend to remove the overlapped args (e.g., `quantization`, `context_window_size`, etc.), shall we set default values for these particular args in `MODEL_PRESETS`? The purpose of introducing `MLC_PRESETS` was to have nice defaults (for each model arch) for such arguments while keeping the `MODEL_PRESETS` (which I renamed to `HF_PRESETS`), but perhaps we can reduce the amount of metadata. Therefore, my main question is whether it would make more sense to add these defaults to `MODEL_PRESETS` or to each `ModelConfig` class?

junrushao commented:
@davidpissarra Apologize for the delay! Let's discuss in more details!

junrushao commented:
> The purpose of introducing MLC_PRESETS was to have nice defaults (for each model arch) for such arguments while keeping the MODEL_PRESETS (which I renamed to HF_PRESETS), but perhaps we can reduce the amount of metadata

I'm reading the `MLC_PRESETS` introduced in this PR, and it seems a set of nice builtin `mlc-chat-config.json` files. Did I get it right? If so, indeed, it's great to have them somewhere approachable so that users could access, and the primary difference is where the best place could be, and there are a couple of solutions:

- A1. Builtins in MLC codebase. Note that some hyperparameters could vary model by model (weight), rather than arch by arch. More specifically, two Llama2-7B models could share exactly the same architecture, but differ only via conversation template - It means we will store all supported models in the codebase, which could be as many as hundreds of them.
- A2. Download `mlc-chat-config.json` from an MLC format HF repo. Note that it means it requires network connection to be available whenever we wanted to compile for a model. It means to compile a model, one has to download it from HF.

> Since we intend to remove the overlapped args (e.g., `quantization`, `context_window_size`, etc.), shall we set default values for these particular args in `MODEL_PRESETS`

Yeah one really nice thing we did in the past few weeks is that we removed almost all duplicated flags between `mlc_chat compile` and `mlc_chat gen_config`, and the next couple of things we want to enable are:

- G1. Making `mlc_chat compile` work on any MLC format repo or directory

```bash
# URL: https://huggingface.co/junrushao/Llama-2-7b-hf-q4f16_1
mlc_chat compile HF://junrushao/Llama-2-7b-hf-q4f16_1
# Local directory
mlc_chat compile /path/to/Llama-2-7b-hf-q4f16_1
```

- G2. Making `ChatModule` work without `model_lib` readily present by just-in-time compilation calling `mlc_chat compile`:

```python
# URL: https://huggingface.co/junrushao/Llama-2-7b-hf-q4f16_1
cm = ChatModule("HF://junrushao/Llama-2-7b-hf-q4f16_1")
# Local directory
cm = ChatModule("/path/to/Llama-2-7b-hf-q4f16_1")
```

- G3. Making `ChatModule` work by searching `model_lib` locally (AOT). This logic exists already, and for now, it searches by exact match of string names. One thing to potentially think about is: can we improve it by searching for model architecture? I believe the answer is indeed yes, and we will need a way to conveniently organize to make sure 1) the libraries are discoverable for models sharing the same architecture; 2) it's convenient enough for us maintainers to release those model_libs

junrushao commented:
Then moving to G3(a), it could actually be formulated as the following question: given a `.so` is exported from an `nn.Module` defined by the command line tool below:

```bash
mlc_chat compile mlc-chat-config.json --extra-flags
```

ignoring some minor differences (e.g. SM architectures used in cuda fatbin, non-default optimization flags), what are the minimal set of flags that uniquely defines a `.so` file, and how to conveniently define it? My answer is: `model_config` + `quantization`.

The disadvantage of this approach is that it could lead to some flags with the same name in `mlc-chat-config.json`, e.g. `config["context_window_size"]` and `config["model_config"]["context_window_size"]`. And actually, the two flags are legal to differ from our cpp runtime's perspective - it's okay for the users specify a smaller `context_window_size` when using the runtime

davidpissarra commented:
Hi @junrushao, my only concern about removing `MLC_PRESETS` was mainly because of the quantization (which is not defined when using a preset), since we removed the overlapping flag `--quantization` on #1371. However, I believe it makes sense to keep this flag in case one uses a preset. As a result, I removed `MLC_PRESETS` and reverted `HF_PRESETS` to `MODEL_PRESETS`, and `detect_quantization` will make sure a quantization scheme is provided either as an argument or inside the config file:

```python
def detect_quantization(quantization_arg: str, config: Path) -> Quantization:
```

Let me know if this looks good, so we can merge it. Thanks a lot!

junrushao commented:
Good idea! With `--quantization` as an exception, we are now able to 100% control the compilation process pretty easily. Let's get this PR in!
